src/main/java/core/basesyntax/service/strategy/strategyimpl/PurchaseOperation.java

```diff
@@ -0,0 +1,18 @@
+package core.basesyntax.service.strategy.strategyimpl;
+
+import core.basesyntax.dao.Storage;
+import core.basesyntax.model.FruitRecord;
+
+public class PurchaseOperation implements OperationHandler {
+    @Override
+    public void apply(FruitRecord transaction) {
+        Integer currentQuantity = Storage.storage.get(transaction.getFruit());
+        int purchaseQuantity = transaction.getQuantity();
+        if (currentQuantity == null && currentQuantity < 0) {
+            throw new IllegalArgumentException("Invalid quantity of for purchase.");
+        }
+        int newQuantity = currentQuantity - purchaseQuantity;
+        Storage.storage.put(transaction.getFruit(), newQuantity);
```

Comment:
The logic in lines 11-15 is a critical part of the operation and it should be encapsulated in a more robust method. Also, consider following the SOLID principles by making this class single-responsible for handling the 'purchase' operation only and not directly interacting with the Storage class.